EIPS/eip-7702.md

```diff
@@ -88,11 +155,33 @@ Specifically:
 
 ## Backwards Compatibility
```

Comment:
Missing from the list of broken assumptions is signature malleability: 
Specifically, a user might sign an unverified experimental 7702 contract on a testnet under the assumption (which was correct until now) that it is a sandbox, and nothing affects other network.
This becomes disastrous, as this contract can be used anywhere - and currently is un-revocable.
The user never opt-in to this new feature: he merely continued his work with an existing (though updated) wallet.

Reply:
@drortirosh chainId variable will be added to the signature. Also, this type of signed data has a specific magic. Wallets are expected to warn the user heavily about what they are about to sign.

Reply:
Signing an unverified / experimental 7702 contract would never happen with a wallet that correctly implements 7702.

Reply:
I don't have any problem with "wallet that correctly implements the standard".

Leaving the chainid out of signature EIP breaks the invariant that "testnet are sandbox and can't affect any other network".
Please add that text into your security consideration, if you think that's acceptable.

Reply:
> Signing an unverified / experimental 7702 contract would never happen with a wallet that correctly implements 7702.

The only way to implement this is by completely blocking (by the wallet) from ever signing anything un-approved by the wallet, not even with a warning message.
That's the only way you can block this concern.

Even with a warning message such as:

> Beware of this signature: it opens your account to malicious operations not only on this network, but on any other network, and this operation is un-revocable. are you sure you want to continue

And I'm afraid that on testnets, some people will STILL sign this message, as the current narrative is "testnets are sandbox"

Reply:
It would be a nice invariant to have, but it's not our current reality. The safety we have relies on apps and wallets to implement things in a safe manner.

At the protocol level, chain ID is already optional. If the wallets correctly implement chain ID by encoding it in the signature, it can safe. But there is nothing forcing them to

For applications and smart contract wallets, anything goes. There are recommendations, such as 
in permit. It is *recommended* implementors include chain id in the domain separator for 712. If tokens implement this correctly, users are okay.

Smart contract wallets are similar. 4337 recommends wallets sign over the chain id, yet it isn't enforced.

712 is a good example by itself. It is used by many on-chain applications for authorization. Chain id is an optional field enforced only at the application level. Yet most users of 712 do sign over the chain id

We can't from the protocol up ensure safety for users. The best we can do is build a robust platform for developers to build safe tools. If the layers above do not uphold their responsibility then the number of ways users will be taken advantage of will be innumerable.

The point is that signing over the chain id is important, but enforcing it at the protocol level is not. Applications and wallet devs have clearly shown they are capable of enforcing it in the signatures they process.

Reply:
> At the protocol level, chain ID is already optional. If the wallets correctly implement chain ID by encoding it in the signature, it can safe. But there is nothing forcing them to

After EIP-155, chainid is **mandatory** - that is, it is optional to sign, but the chain itself (most if not all evm-based chains today) enforce it at the protocol level.

> Smart contract wallets are similar. 4337 recommends wallets sign over the chain id, yet it isn't enforced.

ERC-4337 passes **userOpHash** to the account to sign, which already embed the **chainid** inside (wallet implementation MAY ignore it, and sign whatever it likes (or even not sign at all) - but it is important that no **off-chain** entity can decide, only the smart contract itself)

> We can't from the protocol up ensure safety for users. 

We can't **ensure** - but we should **reduce** the risk as much as possible.

Reply:
EIP-155 is not mandatory.

> The currently existing signature scheme using v = 27 and v = 28 remains valid and continues to operate under the same rules as it did previously.

If your wallet doesn't support it correctly, you could unknowingly sign away your assets on another chain.

Reply:
> EIP-155 chain id is not mandatory.
Is it? Can you submit a non-EIP-155 transaction on mainnet? (or any other chain, for that matter?)
Can you deploy deterministically [Nick's determinisitic deployer](https://github.com/Arachnid/deterministic-deployment-proxy) on new networks? 
All new networks block such "malleable transactions" - and rightly so.
And in order to have this deployer on the same address across networks, they have to submit a "nonstandard state change" (e.g. a fork) to allow this specific non-EIP-155 transaction get through.

This is because malleable transactions are dangerous, and we no longer trust applications or wallets to take this decision.

Reply:
Which networks are banning chain id 0?

Reply:
> Which networks are banning chain id 0?

The question is, on which network you can submit a non-EIP-155 transaction? All block builders on all networks I know of block them.
That's a de-facto standard (that all TXs are EIP-155).

Reply:
I'm saying that there is an implicit assumptions today by users, that this EIP breaks:
1. That signing data on testnet stays in testnet and doesn't affect other networks.
2. That signing anything has a one-shot effect, not long-lasting (or permanent)

And I think that breaking these assumptions AT LEAST deserves being mentioned in the "security consideration" (or backward compatibility) section.
After specifying them, you can add your argument why you think it is OK to break them - but at least acknowledge that we're introducing a network change that alters existing users assumptions.
This is not a change that affect new users that use those new features, or new wallets: it affect all existing users that happen to upgrade their existing wallet, and are unaware of such changes.

Reply:
> That signing data on testnet stays in testnet and doesn't affect other networks.

Again, this is not correct. The protocol allows for non EIP-155 txs to be replayed on any chain. If users think they're safe because the wallet says they're on a testnet today, it's because apps, wallets, and RPC providers have done well to create this feeling of security. They could have just as easily continued signing and allowing txs w/o chain id baked in. I have no reason to believe post-7702 things will be any different.

> That signing anything has a one-shot effect, not long-lasting (or permanent)

This would be true for 7702 auths. Happy to put reword the security considerations if it is not clear there.